### ISSUE TITLE
Detailed diff output 

### ISSUE THREAD

**Issue by skorfmann** [hashicorp/terraform-cdk#114]
Currently it's possible to get a quick overview of which resources will be change. However, currently there's no way to dive deeper into the the actual changes.

(Daniel's edit)
For the 0.7 release, this will be minimally scoped to display the Terraform Core diff output, in order to show the needed information without requiring a major design effort. In the future, pending user feedback, we expect to do more design to find a diff representation that's more idiomatic to CDK for Terraform users.

(Sebastian's original issue)
I think we'll have to have two different modes: 

### Interactive 

Let the user navigate through the list of resources and show a panel on the side with the actual diff related tot the currently selected resource.

### Non Interactive (CI) 

Just show the the entire, detailed diff for inspection in automated environments.

**Comment by kaisellgren** [comment]
I hope we can get better output for CI environments, because this is what we have now (unless there's a feature I'm not aware of):

![image](https://user-images.githubusercontent.com/148256/95620554-0d932600-0a79-11eb-9dbd-85a31fcaca18.png)

**Comment by jsteinich** [comment]
@kaisellgren setting the `CI` environment variable should help with that.

**Comment by DovydasNavickas** [comment]
Getting actual changes while checking diff seems essential, because we had a case of Terraform restarting the provisioned VMs no matter if there were any changes or not.
The invisible change was a cloud init user (`ciuser`) property being set after the VM is provisioned initially to a user that already exists in the VM template, which we only found while inspecting and diffing the debug output of `cdktf diff` with a `CDKTF_LOG_LEVEL` flag. In our CDK source we simply omitted the `ciuser` and thus Terraform tried to revert it to `null` and that was considered as a change needing a restart.

So, we ran this and got the debug output:
```
CDKTF_LOG_LEVEL=debug cdktf diff main-k8s-cluster
```

But after running diff in debug mode, not only the output is minified (`.tfstate` is not), but also has resources listed in 4th level of JSON (`planned_values.root_module.resources`):
```jsonc
{
    // <...>
    "planned_values" : {
        "root_module": {
            "resources": [
                // We found the change in here.
            ]
        }
    }
    // <...>
}
```

And in `.tfstate`, these resources are in the first level:
```jsonc
{
    // <...>
    "resources": [
        // Resources are defined here in .tfstate
    ]
    // <...>
}
```

Thus, diffing becomes manual work copying debug output, beautifying it, taking excerpts of planned value for resources and only then diffing it with the `.tfstate` ones.

Non-CDK Terraform states the changes clearly, what specific keys values have changed, what they were before planning and what did they change to. Thus, it would be great to get the same from `cdktf diff`.

**Comment by jsteinich** [comment]
@DovydasNavickas another (likely easier to read) option is to run `terraform plan` rather than `cdktf diff`. Just make sure to run `cdktf synth` before the plan and to change to the stack output directory before running terraform.

**Comment by simon-enplore** [comment]
Sorry for commenting on an old issue, but is there any progress on this / is there some way to contribute to the progress of this?

**Comment by phinze** [comment]
From discussion:

 * We are thinking we'd like to tackle this in the same release as #650, since that behavior creates things we need to think about as to how we display outputs for more than one stack at once.
 * Because we'd like to get #651 done before that, this may slide one more release back.

Hi @simon-enplore - thanks for checking in on this. We're definitely still planning to improve the diff output; it's just a matter of us figuring out how it fits in with our other work. Since there's more definition work necessary here, it's hard for us to accept contributions on this feature, but thank you for the offer!

We will work to keep this issue updated with our plans as they solidify.

**Comment by DanielMSchmidt** [comment]
@ansgarm Is this a duplicate of #1097 ?

**Comment by ansgarm** [comment]
@DanielMSchmidt almost – this ticket is more like #807 (as in Design / Planning work) and #1097 should be updated with the implementation steps as soon as the design is set (by this ticket).

**Comment by M1kep** [comment]
The original issue here stated that for the .7 release it was going to be targeted to fall back to the `terraform plan` output. In reading this thread I'm not too sure I've seen something holding that back? Some issues I have with the current diff output:

* No detailed information
* Column size appears to be static so the text is quite hard to read even if you only wanted to know _what_ is changing.

I'm currently using the below as workaround via my package.json(for TypeScript/JS at least), I assume this could be instead used via a make file.

package.json
```json
{
  "name": "cdktf-testing",
  "version": "1.0.0",
  "main": "main.js",
  "types": "main.ts",
  "license": "MPL-2.0",
  "private": true,
  "scripts": {
    // Your other scripts
    "plan": "cdktf synth && terraform -chdir=./cdktf.out/stacks/github-infra plan",
    "plan-norefresh": "cdktf synth && terraform -chdir=./cdktf.out/stacks/github-infra plan -refresh=false"
  },
  "engines": {
    "node": ">=10.12"
  },
  "dependencies": {
    // Your dependencies
  },
  "devDependencies": {
    "cdktf-cli": "^0.8.3"
  }
}```

With this you can just run `npm run plan` to get the latest plan

**Comment by DanielMSchmidt** [comment]
Our plan for this is to replace current dynamic view with a stream of logs from terraform that are more terminal-native and show all the context needed. Mid-term we want to add some more cdktf-specific context to the logs like language-native property / resource names and construct paths in the logs.

**Comment by Tristan971** [comment]
While it's understandable to want to plan further ahead w/ multi-stack support, a first step that makes the underlying `terraform plan` output non-debug (or togglable as info rather than debug somewhere in cdktf.json) would probably be a quick win...?

**Comment by DanielMSchmidt** [comment]
Hey @Tristan971, this is pretty much what I am working on right now, please take a look at the video in the PR description: https://github.com/hashicorp/terraform-cdk/pull/1596

**Comment by Tristan971** [comment]
That's great! Thanks a lot @DanielMSchmidt, will keep a close eye on it 👍

**Comment by DanielMSchmidt** [comment]
Closed by #1596 (streaming the logs) and  #1635 (enhancing the logs)
